File: services/server/src/message/message.py
import socket
import safe_socket
import logger #TODO
from lottery.bet import Bet

END_MESSAGE =  b'\x00'
BET_MESSAGE = b'\x01'
BATCH_MESSAGE = b'\x02'

def recive_message(socket: socket):
    type = safe_socket.recv_all(socket, 1)

    if type == BET_MESSAGE:
        return [recive_bet_message(socket)]
    elif type == END_MESSAGE:
        return None
    elif type == BATCH_MESSAGE:
        logger.info( "a", logger.LogResult.success, "messages-amount", type)
        return recibe_batch_message(socket)
    else:
        return None

def recibe_batch_message(socket: socket.socket):
    info = safe_socket.recv_all(socket, 8)
    if len(info) != 8:
        return None

    bets_size = int.from_bytes(info[0:4], byteorder='big')
    bets_lenght = int.from_bytes(info[4:8], byteorder='big')

    bets_bytes = safe_socket.recv_all(socket, bets_lenght)

    if len(bets_bytes) != bets_lenght:
        return None

    bets = [] 
    size = 0

    for i in range(bets_size):
        agency_id = int.from_bytes(bets_bytes[size:size+4], byteorder='big')
        document = int.from_bytes(bets_bytes[size+4:size+8], byteorder='big')
        number = int.from_bytes(bets_bytes[size+8:size+12], byteorder='big')
        
        first_name_len = int.from_bytes(bets_bytes[size+12:size+16], byteorder='big')
        last_name_len = int.from_bytes(bets_bytes[size+16:size+20], byteorder='big')
        birthdate_len = int.from_bytes(bets_bytes[size+20:size+24], byteorder='big')

        size += 24

        first_name = bets_bytes[size:size+first_name_len].decode()
        last_name = bets_bytes[size+first_name_len:size+first_name_len+last_name_len].decode()
        birthdate = bets_bytes[size+first_name_len+last_name_len:size+first_name_len+last_name_len+birthdate_len].decode()

        size += (first_name_len+last_name_len+birthdate_len)
        bets.append(Bet(agency_id, first_name, last_name, document, birthdate, number))

    return bets


def send_batch_message(socket: socket.socket, bets):
    bytes = BATCH_MESSAGE
    bytes += len(bets).to_bytes(4, byteorder='big')

    bet_bytes = b''

    for bet in bets:
        bet_bytes += bet_to_bytes(bet)

    bytes += len(bet_bytes).to_bytes(4, byteorder='big')
    bytes += bet_bytes

    safe_socket.send_all(socket, bytes)

def recive_bet_message(socket: socket.socket):

    integers = safe_socket.recv_all(socket, 24)

    if len(integers) < 24 :
        return None

    agency_id = int.from_bytes(integers[0:4], byteorder='big')
    document = int.from_bytes(integers[4:8], byteorder='big')
    number = int.from_bytes(integers[8:12], byteorder='big')

    first_name_len = int.from_bytes(integers[12:16], byteorder='big')
    last_name_len = int.from_bytes(integers[16:20], byteorder='big')
    birthdate_len = int.from_bytes(integers[20:24], byteorder='big')

    strings = safe_socket.recv_all(socket, first_name_len + last_name_len + birthdate_len)

    first_name = strings[0:first_name_len].decode()
    last_name = strings[first_name_len:first_name_len+last_name_len].decode()
    birthdate = strings[first_name_len+last_name_len:].decode()

    return Bet(agency_id, first_name, last_name, document, birthdate, number)

def send_bet_message(socket: socket.socket, bet):
    bytes = BET_MESSAGE
    bytes += bet_to_bytes(bet)

    safe_socket.send_all(socket, bytes)

def end_bet_message(socket: socket.socket):
    safe_socket.send_all(socket, END_MESSAGE)

def bet_to_bytes(bet):
    bytes = b''

    bytes += bet.agency_id.to_bytes(4, byteorder='big')
    bytes += bet.document.to_bytes(4, byteorder='big')
    bytes += bet.number.to_bytes(4, byteorder='big')

    bytes += len(bet.first_name).to_bytes(4, byteorder='big')
    bytes += len(bet.last_name).to_bytes(4,  byteorder='big')
    bytes += len(bet.birthdate).to_bytes(4, byteorder='big')

    bytes += bet.first_name.encode()
    bytes += bet.last_name.encode()
    bytes += bet.birthdate.encode()

    return bytes

File: services/client/src/message/message.go
package message

import (
	"net"
	"strings"
	"strconv"
	"errors"
	"encoding/binary"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
	//"github.com/7574-sistemas-distribuidos/tp-nivelador/src/logger"
)

const END_MESSAGE byte = 0b00000000
const BET_MESSAGE byte = 0b00000001
const BATCH_MESSAGE byte = 0b00000010

type BetMessage struct{
	Agency_id int
	First_name string
	Last_name string
	Document int
	Birthdate string
	Number int
}

func ReciveMessage(conn net.Conn) ([]BetMessage, error){
	msgType, err := safe_socket.RecvAll(conn, 1)

	if err != nil {
		return []BetMessage{}, err
	}
	
	if msgType[0] == BET_MESSAGE{
		return ReciveBetMessage(conn)
	}else if msgType[0] == END_MESSAGE{
		return []BetMessage{}, nil
	}else if msgType[0] == BATCH_MESSAGE{
		return ReciveBatchMessage(conn)
	}else{
		return []BetMessage{}, nil
	}
}

func CreateMessageBet(agencyIdStr string, message string) (BetMessage, error){

	info := strings.Split(message, ",")

	if len(info) != 5 {
		return BetMessage{}, errors.New("mensaje inválido")
	}

	agencyId, err := strconv.Atoi(agencyIdStr)
	if err != nil {
		return BetMessage{}, err
	}
	document, err := strconv.Atoi(info[2])
	if err != nil {
		return BetMessage{}, err
	}
	number, err := strconv.Atoi(info[4])
	if err != nil {
		return BetMessage{}, err
	}
	
	betMesage := BetMessage{agencyId, info[0], info[1], document, info[3], number}

	return betMesage, nil
}

func ReciveBatchMessage(conn net.Conn) ([]BetMessage, error){
	info, err := safe_socket.RecvAll(conn, 8)

	if len(info) < 8 || err != nil{
		return []BetMessage{}, errors.New("mensaje inválido")
	}

	betsSize := int32(binary.BigEndian.Uint32(info[0:4]))
	betsLenght := int32(binary.BigEndian.Uint32(info[4:8]))

	betsBytes, err := safe_socket.RecvAll(conn, int(betsLenght))
	if err != nil || len(betsBytes) < int(betsLenght) {
		return []BetMessage{}, errors.New("mensaje inválido")
	}

	bets := []BetMessage{}
	size := 0
	for i := 0; i < int(betsSize); i++ {
		agencyId := int32(binary.BigEndian.Uint32(betsBytes[size:size+4]))
		document := int32(binary.BigEndian.Uint32(betsBytes[size+4:size+8]))
		number := int32(binary.BigEndian.Uint32(betsBytes[size+8:size+12]))

		first_name_len := int32(binary.BigEndian.Uint32(betsBytes[size+12:size+16]))
		last_name_len := int32(binary.BigEndian.Uint32(betsBytes[size+16:size+20]))
		birthdate_len :=  int32(binary.BigEndian.Uint32(betsBytes[size+20:size+24]))

		size += 24

		first_name := string(betsBytes[size:first_name_len])
		last_name := string(betsBytes[size+int(first_name_len):size+int(first_name_len+last_name_len)])
		birthdate := string(betsBytes[size+int(first_name_len+last_name_len):size+int(first_name_len+last_name_len+birthdate_len)])

		size += int(first_name_len+last_name_len+birthdate_len+birthdate_len)

		bets = append(bets, BetMessage{int(agencyId), first_name, last_name, int(document), birthdate, int(number)})
	}	
	
	return bets, nil
}

func SendBatchMessage(conn net.Conn, bets []BetMessage) error{
	var bytes []byte

	bytes = append(bytes, []byte{byte(BATCH_MESSAGE)}...)

	buf := make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(len(bets)))
	bytes = append(bytes, buf...)

	var bet_bytes []byte

	for _, bet := range bets {
		bet_bytes = append(bet_bytes, betToBytes(bet)...)
	}

	buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(len(bet_bytes)))
	bytes = append(bytes, buf...)

	bytes = append(bytes, bet_bytes...)
	
	if err := safe_socket.SendAll(conn, bytes); err != nil {
		return err
	}

	return nil
}

func ReciveBetMessage(conn  net.Conn) ([]BetMessage, error){
	integers, err := safe_socket.RecvAll(conn, 24)
	if len(integers) < 24 || err != nil{
		return []BetMessage{}, errors.New("mensaje inválido")
	}

	agencyId := int32(binary.BigEndian.Uint32(integers[0:4]))
	document := int32(binary.BigEndian.Uint32(integers[4:8]))
	number := int32(binary.BigEndian.Uint32(integers[8:12]))

	first_name_len := int32(binary.BigEndian.Uint32(integers[12:16]))
	last_name_len := int32(binary.BigEndian.Uint32(integers[16:20]))
	birthdate_len :=  int32(binary.BigEndian.Uint32(integers[20:24]))

	strings, err := safe_socket.RecvAll(conn, int(first_name_len + last_name_len + birthdate_len))
	if err != nil{
		return []BetMessage{}, errors.New("mensaje inválido")
	}

	first_name := string(strings[0:first_name_len])
	last_name := string(strings[first_name_len:first_name_len+last_name_len])
	birthdate := string(strings[first_name_len+last_name_len:first_name_len+last_name_len+birthdate_len])

	return []BetMessage{{int(agencyId), first_name, last_name, int(document), birthdate, int(number)}}, nil
}

func SendBetMessage(conn net.Conn, betMesage BetMessage) error{

	var bytes []byte

	bytes = append(bytes, []byte{byte(BET_MESSAGE)}...)
	bytes = append(bytes, []byte(betToBytes(betMesage))...)

	if err := safe_socket.SendAll(conn, bytes); err != nil {
		return err
	}

	return nil
}

func EndBetMessages(conn net.Conn){
	safe_socket.SendAll(conn, []byte{byte(END_MESSAGE)})
}

func betToBytes(betMesage BetMessage) []byte{
	var bytes []byte

	buf := make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(betMesage.Agency_id))
	bytes = append(bytes, buf...)

	buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(betMesage.Document))
	bytes = append(bytes, buf...)

	buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(betMesage.Number))
	bytes = append(bytes, buf...)
	
	buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(len(betMesage.First_name)))
	bytes = append(bytes, buf...)

	buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(len(betMesage.Last_name)))
	bytes = append(bytes, buf...)

	buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(len(betMesage.Birthdate)))
	bytes = append(bytes, buf...)

	bytes = append(bytes, []byte(betMesage.First_name)...)
	bytes = append(bytes, []byte(betMesage.Last_name)...)
	bytes = append(bytes, []byte(betMesage.Birthdate)...)

	return bytes
}